src/Microsoft.VisualStudio.Threading/AsyncLazy.cs

```diff
@@ -147,19 +147,55 @@ public Task<T> GetValueAsync(CancellationToken cancellationToken)
                         resumableAwaiter = new InlineResumable();
                         Func<Task<T>> originalValueFactory = this.valueFactory;
                         this.valueFactory = null;
-                        Func<Task<T>> valueFactory = async delegate
+
+                        Task<T> ValueFactory()
                         {
-                            try
-                            {
-                                await resumableAwaiter;
-                                return await originalValueFactory().ConfigureAwait(continueOnCapturedContext: false);
-                            }
-                            finally
+                            var taskCompletionSource = new TaskCompletionSource<T>();
+                            async Task InnerValueFactory()
```

Comment:
This would seem to be creating yet another closure (allocating another object). Is that required to solve the problem, or can we do some more work to avoid additional GC pressure from this change?